docs/examples/use_cases/tensorflow/yolov4/src/dali/ops.py

```diff
@@ -0,0 +1,158 @@
+        bboxes,
+        M=[0.5,  0.0,  0.5,  0.0,
+           0.0,  0.5,  0.0,  0.5,
+          -1.0,  0.0,  1.0,  0.0,
+           0.0, -1.0,  0.0,  1.0]
+    )
+
+# Transforms bbox ltrb coordinates, so that they fit within a window
+#   anchored at (pos_x, pos_y) and with a shape (shape_x, shape_y).
+def bbox_adjust_ltrb(bboxes, shape_x, shape_y, pos_x, pos_y):
+    sx, sy, ex, ey = pos_x, pos_y, shape_x + pos_x, shape_y + pos_y
+    MT = dali.fn.transforms.crop(
+        to_start=dali.fn.stack(sx, sy, sx, sy),
+        to_end=dali.fn.stack(ex, ey, ex, ey)
+    )
+    return dali.fn.coord_transform(bboxes, MT=MT)
+
+# Selects if_true or if_false tensor based on the predicate value.
+# predicate should take integer value equal to either 0 or 1.
+# if_true and if_false should be 2D tensors with equal size along axis 1.
+def select(predicate, if_true, if_false):
+    true_shape = dali.fn.cast(dali.fn.shapes(if_true), dtype=dali.types.DALIDataType.INT32)
+    false_shape = dali.fn.cast(dali.fn.shapes(if_false), dtype=dali.types.DALIDataType.INT32)
+
+    joined = dali.fn.cat(if_true, if_false)
+    sh = predicate * true_shape + (1 - predicate) * false_shape
+
+    st = dali.fn.cat(
+            dali.fn.slice(true_shape * (1 - predicate), start=[0], shape=[1], axes=[0]),
+            dali.fn.constant(idata=0, shape=[1])
+        )
+
+    return dali.fn.slice(joined, start=st, shape=sh, axes=[0,1], out_of_bounds_policy="trim_to_shape")
+
+# Based on https://github.com/AlexeyAB/darknet/blob/005513a9db14878579adfbb61083962c99bb0a89/src/image.c#L1297
+def color_twist(images):
+    def random_value():
+        value = dali.fn.random.uniform(range=(1, 1.5))
+        coin = dali.fn.random.coin_flip()
+        return coin * value + (1.0 - coin) * (1.0 / value)
+
+    return dali.fn.color_twist(images,
+        hue=dali.fn.random.uniform(range=(-18.0, 18.0)),
+        brightness=random_value(),
+        contrast=random_value()
+    )
+
+def flip(images, bboxes):
+    coin = dali.fn.random.coin_flip()
+    images = dali.fn.flip(images, horizontal=coin)
+    bboxes = dali.fn.bb_flip(bboxes, horizontal=coin, ltrb=True)
+    return images, bboxes
+
+
+# Performs mosaic using MultiPaste operator.
+def mosaic(images, bboxes, labels, image_size):
+    def generate_tiles(bboxes, labels, shape_x, shape_y):
+        idx = dali.fn.batch_permutation()
+        permuted_boxes = dali.fn.permute_batch(bboxes, indices=idx)
+        permuted_labels = dali.fn.permute_batch(labels, indices=idx)
+        shape = dali.fn.stack(shape_y, shape_x)
+        in_anchor, in_shape, bbx, lbl = dali.fn.random_bbox_crop(
+            permuted_boxes,
+            permuted_labels,
+            input_shape=image_size,
+            crop_shape=shape,
+            shape_layout="HW",
+            allow_no_crop=False,
+            total_num_attempts=64
+        )
+
+        in_anchor = dali.fn.stack(dali.fn.reductions.sum(in_anchor), dali.fn.reductions.sum(in_anchor)) - in_anchor
```

Comment:
What does it do?
It looks like you have `in_anchor = [x, y]` and the result of this line is:
```[x + y - x, x + y - y] = [y, x]``` - basically swapping coordinates! I think that there are a lot of simpler ways to achieve this.

Even assuming that the trick with a sum is the way to go, calling reduction on the same object twice is quite wasteful. The stacking is also unnecessary, as we can broadcast scalars in arithm ops:

```suggestion
        in_anchor = dali.fn.reductions.sum(in_anchor) - in_anchor
```
Arguably a more readable option would be:
```
        in_anchor = dali.fn.coord_transform(in_anchor, M = [0, 1,
                                                 1, 0])
```

Reply:
Simply coord_transform won't work, because it requires tensor to be 2d - it would have to be wrapped in expand_dims + squeeze. I went with the first suggestion, as it's clearer in my opinion.